### PR TITLE
Optimisation

### PR DESCRIPTION
**Pull request by Jackojc**
- Use string_views
- New alias for shared_ptr to std::string
- Turn off syncing with stdio for receive function
- Fix cast warnings

### CODE REVIEW

File: module.h [hunk 1]
```diff
@@ -61,38 +62,59 @@ namespace ws::module {
...
     }
 
 
 
     // Accumulate all data and return a single string.
-    inline std::string receive_all(std::streamsize buffer_size) {
-        std::string accumulator;
+    inline Str receive_all(std::streamsize buffer_size) {
+        Str accumulator = std::shared_ptr<std::string>(new std::string());
+        accumulator->reserve(static_cast<unsigned long>(buffer_size));
```
Comment (Repertoi-e) [inline module.h]: This doesn't actually reserve enough space, since there can be any number of `buffer_size`sized buffers.

Comment (Jackojc) [reply module.h]: It reserves enough for if receive is only called once, I didn't want to make it super huge.

Comment (Repertoi-e) [reply module.h]: Ah, sure.

File: module.h [hunk 2]
```diff
@@ -61,38 +62,59 @@ namespace ws::module {
...
         std::string buffer(static_cast<uintmax_t>(buffer_size), '\0');
         int packet_id = 0;
 
-
         while (std::cin.read(&buffer[0], buffer_size)) {
-            callback(buffer, packet_id, false);
+            callback(
+                std::string_view{
+                    buffer.c_str(),
+                    static_cast<unsigned long>(buffer_size)
```
Comment (Hazurl) [inline module.h]: use std::size_t

File: module.h [hunk 3]
```diff
@@ -61,38 +62,59 @@ namespace ws::module {
...
             ++packet_id;
         }
 
 
-        buffer.erase(buffer.begin() + std::cin.gcount(), buffer.end());
-        callback(buffer, packet_id, true);
+        callback(
+            std::string_view{
+                buffer.c_str(),
+                static_cast<unsigned long>(std::cin.gcount())
```
Comment (Hazurl) [inline module.h]: std::size_t

File: module.h [hunk 4]
```diff
@@ -61,38 +62,59 @@ namespace ws::module {
 
 
 
+    using Str = std::shared_ptr<std::string>;
```
Comment (Hazurl) [inline module.h]: Why not std::unique_ptr ? and even, why not std::string ? the move constructor is fast...

File: token.h [hunk 5]
```diff
@@ -1,60 +1,88 @@
...
         std::size_t size;
         if (!read(is, &size, sizeof(size)))
             return is;
 
-        token.content.resize(size);
-        read(is, token.content.data(), size);
 
-        return is;
-    }
+        std::string content(token.content);
```
Comment (Hazurl) [inline token.h]: Did you test the function ? because you don't read into token.content I think...

File: token.h [hunk 6]
```diff
@@ -1,60 +1,88 @@
...
+namespace ws::token {
     using type_t = std::uint16_t;
 
 
     struct Token {
         type_t type;
         std::uint32_t line;
         std::uint32_t column;
-        std::string content;
+        std::string_view content;
```
Comment (Hazurl) [inline token.h]: I though about that, but std::string_view doesn't own the string, which means you have to keep the string alive until content goes out of scope, but the string is read from the input stream which is not guaranteed to be alive as you wants...

Comment (Hazurl) [reply token.h]: string_view can be use serialize the token tho, but not for deserialize the token I think

File: module.h (hunk 2 above)
Comment (Jackojc) [reply module.h]: I think the most correct approach is something like: `std::string_view::size_type` but it seems to not like that :/

Comment (Hazurl) [reply module.h]: size_type is guaranteed to be std::size_t

File: token.h (hunk 6 above)
Comment (Jackojc) [reply token.h]: That's true, yeah

Comment (Jackojc) [reply token.h]: If only JSON was still feasible :(

**Comment by Jackojc** [comment]
I have reverted these changes in the latest PR because they only help marginally and add needless complexity to the code.

**Pull request closed without merging**